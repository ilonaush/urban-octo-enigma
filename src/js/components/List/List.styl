@import "~js/vars"

table
  background-color white
  width 100%
  padding 15px
  tr
    border-bottom 1px solid $gray
  th
    color $orange
    font-size 20px
    font-weight 300
    font-family $montserrat-thin
    text-align left
    &.time
      width 20%
