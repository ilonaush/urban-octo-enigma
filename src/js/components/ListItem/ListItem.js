import React, {Component} from 'react';
import PropTypes from 'prop-types';
import "./ListItem.styl";
import MaskedInput from 'react-text-mask'

class ListItem extends Component {
     constructor(props) {
         super(props);
         this.state = {
             editing: ''
         };
         this.editTime = this.editTime.bind(this);
         // this.handleChange = this.handleChange.bind(this);
         this.saveTime = this.saveTime.bind(this);
     }

    editTime(e) {
        this.setState({
            editing: e.currentTarget.id
        })
    }

    saveTime({target}) {
        this.setState({
            editing: ''
        });
        const worker = {
            ...this.props.worker,
            [target.name]: target.value
        };
        this.props.editWorkerTime(worker);
    }

    render() {
     const {editing} = this.state;
     const {worker: {fullname, position, arrival = '', leaving = ''} = {}} = this.props;
        return (
            <tr>
                <td>
                    {fullname}
                </td>
                <td>
                    {position}
                </td>
                <td id='arrival' onClick={this.editTime} onBlur={this.saveTime} className='worker-time'>
                    {editing === 'arrival' ? <MaskedInput
                        mask={[' ', /\d/,  /\d/, ':',  /\d/,  /\d/]}
                        className='time-input'
                        name='arrival'
                        guide={true}
                        // onChange={this.handleChange}
                        placeholder='––:––'
                    />
                    : arrival ? arrival : 'Click to select'}
                </td>
                <td id='leaving' onClick={this.editTime} onBlur={this.saveTime} className='worker-time'>
                    {editing === 'leaving' ? <MaskedInput
                        mask={[' ', /\d/,  /\d/, ':',  /\d/,  /\d/]}
                        className='time-input'
                        name='leaving'
                        guide={true}
                        // onChange={this.handleChange}
                        placeholder='––:––'
                    />
                    : leaving ? leaving : 'Click to select'}
                </td>
            </tr>
        );
    }
}

ListItem.propTypes = {};

export default ListItem;
