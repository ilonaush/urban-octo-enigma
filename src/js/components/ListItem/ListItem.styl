.worker-time
  cursor pointer

.time-input
    width 50px
    margin 0 auto
