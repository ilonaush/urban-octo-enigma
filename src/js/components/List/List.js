import React, {Component} from 'react';
import PropTypes from 'prop-types';
import "./List.styl";
import {bindActionCreators} from "redux";
import actions from "../../reducers/actions";
import {connect} from "react-redux";
import ListItem from "../ListItem/ListItem";
import {Link} from "react-router-dom";

class List extends Component {


    render() {
        return (
            this.props.workers && this.props.workers.length ?
            <table>
                <thead>
                    <tr>
                        <th>
                            Name
                        </th>
                        <th>
                            Position
                        </th>
                        <th className='time'>
                            Arrival
                        </th>
                        <th className='time'>
                            Leaving
                        </th>
                    </tr>
                </thead>
                <tbody>
                {this.props.workers.map ((worker) =>
                    <ListItem key={worker.id} worker={worker} editWorkerTime={this.props.actions.editWorkTime}/>
                 )}
                </tbody>
            </table> : <Link to='/add-worker'>Add worker</Link>
        );
    }
}

List.propTypes = {};

function mapStateToProps(state) {
    return { workers: state.workers  }
}
function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(actions, dispatch)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(List);
