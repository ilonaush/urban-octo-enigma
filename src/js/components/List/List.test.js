import React from 'react';
import List  from './List';
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import renderer from "react-test-renderer";
import { mount, shallow } from 'enzyme';
import ListItem from "../ListItem/ListItem";
import configureMockStore from "redux-mock-store";


const mockStore = configureMockStore();
const store = mockStore({ cats: [
        {id: 1, fullname: 'mike', position: 'cat'},
        {id: 2, fullname: 'mike', position: 'cat'},
        {id: 3, fullname: 'mike', position: 'cat'},
        {id: 4, fullname: 'mike', position: 'cat'}
        ]
});
const emptyStore = mockStore({});

describe('List', function () {

    beforeEach(() => {
        this.wrapper = mount(
            <MemoryRouter keyLength={0}>
                <Provider store={store}>
                    <List/>
                </Provider>
            </MemoryRouter>
        );
    });
    it('renders', () => {
        expect(this.wrapper.find(List)).toMatchSnapshot();
    });

    it('renders as many rows as expected', () => {

        const ListComponent = mount(
            <MemoryRouter>
                <Provider store={store}>
                    <List/>
                </Provider>
            </MemoryRouter>
        );
        expect(ListComponent.find(ListItem).length).toEqual(4);
    });
    it('renders no row when store is empty', () => {

        const ListComponent = mount(
            <MemoryRouter>
                <Provider store={emptyStore}>
                    <List/>
                </Provider>
            </MemoryRouter>
        );
        expect(ListComponent.find(ListItem).length).toEqual(0);
    })
    it('renders button when store is empty', () => {
        const ListComponent = mount(
            <MemoryRouter>
                <Provider store={emptyStore}>
                    <List/>
                </Provider>
            </MemoryRouter>
        );
        expect(ListComponent.find('.home-add-btn').exists()).toEqual(true);
    })
});
