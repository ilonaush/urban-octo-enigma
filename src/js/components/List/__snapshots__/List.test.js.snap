// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Position
      </th>
      <th
        className="time"
      >
        Arrival
      </th>
      <th
        className="time"
      >
        Leaving
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        mike
      </td>
      <td>
        cat
      </td>
      <td
        className="worker-time"
        id="arrival"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
      <td
        className="worker-time"
        id="leaving"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
    </tr>
    <tr>
      <td>
        mike
      </td>
      <td>
        cat
      </td>
      <td
        className="worker-time"
        id="arrival"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
      <td
        className="worker-time"
        id="leaving"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
    </tr>
    <tr>
      <td>
        mike
      </td>
      <td>
        cat
      </td>
      <td
        className="worker-time"
        id="arrival"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
      <td
        className="worker-time"
        id="leaving"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
    </tr>
    <tr>
      <td>
        mike
      </td>
      <td>
        cat
      </td>
      <td
        className="worker-time"
        id="arrival"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
      <td
        className="worker-time"
        id="leaving"
        onBlur={[Function]}
        onClick={[Function]}
      >
        Click to select
      </td>
    </tr>
  </tbody>
</table>
`;
